rpctest: Add debug and trace facility to rpctest.

marcopeereboom opened the pull request:
The rpctest suite is useful however it is impossible to debug in its
current state. This diff adds debug and tracing facilities to the test
that are worth keeping.

It adds two global flags `debug` and `trace` that can be enabled when
additional output is required. In addition, it sends the test object
around so that segregated parts of the code use the same logging
facility. And lastly, it will always print `dcrd` panics to the console.

davecgh commented on rpctest/node.go:
```diff
@@ -180,46 +211,115 @@ func newNode(config *nodeConfig, dataDir string) *node {
+				n.tracef("stderr: EOF")
+				return
+			}
+			// It's ok to race n.pid.
```
Races are actually never ok because they can be used access arbitrary memory.

marcopeereboom replied:
It really is in this case but I'll perform some magic. This will require a really crappy fix.

We'll need to launch the process WITHOUT logging in the pipe readers.

davecgh replied:
It doesn't directly apply to this case, but a good article that showcases one way that a race can be used to access arbitrary memory is https://research.swtch.com/gorace.

Also, you can check the official Go memory model documentation at https://golang.org/ref/mem (courtesy of @jrick).

marcopeereboom replied:
I added a waitgroup to handle this.

I am pretty familiar with both those links :-)

marcopeereboom commented on rpctest/rpc_harness.go:
```diff
@@ -146,11 +149,14 @@ func New(activeNet *chaincfg.Params, handlers *rpcclient.NotificationHandlers, e
 	}
 
+	// Uncomment and change to enable additional dcrd debug/trace output.
+	//config.debugLevel = "MINR=trace,TXMP=trace,TRSY=trace,RPCS=trace,PEER=trace"
```
It's disabled. Reason I left it there is as an example.

davecgh replied:
I know.  I was saying that if you uncomment this currently to use it, it will break because there is no `TRSY` subsystem yet.

marcopeereboom commented on rpctest/debug.go:
```diff
@@ -0,0 +1,35 @@
+// Copyright (c) 2016-2017 The btcsuite developers
```
You always include the old license in perpetuity.

jrick replied:
I'm nitpicking here, but this isn't the license. Only the full license text (which we only keep in the repo root, not duplicated in each file) is the actual license.

marcopeereboom commented:
I added a crash dumper when nodes hang.

That said, I think this test should be removed. It is irredeemably broken. It mixes OO ideas with global accessors in objects with multiple locks. It also uses a map of pointers that are by definition racing (running processes) and on top of all that it mixes several mutexes in various call orders.

I love what the test attempts to do but you never know if you are debugging the test or the code. It should be removed and rewritten.
